src/core/document.js

```diff
@@ -446,6 +472,23 @@ class Page {
       newAnnotationsPromise,
     ]).then(function ([pageOpList, annotations, newAnnotations]) {
       if (newAnnotations) {
+        // Some annotations can already exist (if it has the refToReplace
+        // property). In this case, we replace the old annotation by the new
+        // one.
+        for (let i = 0, ii = newAnnotations.length; i < ii; i++) {
+          const newAnnotation = newAnnotations[i];
+          if (newAnnotation.refToReplace) {
+            const j = annotations.findIndex(
+              a => a.ref === newAnnotation.refToReplace
+            );
+            if (j >= 0) {
+              annotations.splice(j, 1, newAnnotation);
```

Comment:
Given how Objects, remember that Arrays are essentially a special-case of those, are passed by reference in JavaScript this will modify the cached annotations-data and affect future `getAnnotations`-calls in a way that I cannot imagine is correct/intended here.

Fixing this ought to mean simply adding `annotations = annotations.slice();` e.g. [before this line](https://github.com/mozilla/pdf.js/pull/16523/files#diff-f0385d138acfe65f04aec809ad0522722883e56b6420ccb8f1d2e16b50631c1eR475), another *possible* solution might be to just place [this new line](https://github.com/mozilla/pdf.js/pull/16523/files#diff-f0385d138acfe65f04aec809ad0522722883e56b6420ccb8f1d2e16b50631c1eR491) at that spot instead.